rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,501 @@
+
+Another possibility is that the Kube Agent might use the invite token each time the pod is recycled, meaning that it issues a join request to the cluster every time it starts and receives a new identity from the Auth service. This means that the invite token must be a static/long-lived token, otherwise after a while the agent could not register himself in the cluster because the invite token expired. This approach is not recommended and might cause security flaws because the join token can be stolen or guessed.
+
+One solution that might address all the issues referenced above is to allow the Kube agent to use Kubernetes secrets as storage backend. This allows not only that the agent is able to run stateless depending only on native objects generally available in any Kubernetes cluster, but also that the agent might support dynamic short-lived invite tokens with no dependency on external storage.
+
+## Details
+
+### Limitations
+
+High availability can only be achieved using Statefulsets and not deployments. This means that if the desired number of replicas is bigger than one, i.e `replicas>1`, the [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) has to switch to Statefulset objects instead of Deployments. This change is required because some Kubernetes invariant must be kept between restarts in order to correctly map each agent pod to its identity stored in the secret. The invariant used is the Statefulset pod names `{{ .Release.Name }}-{0...replicas}}`.
+
+Given this, it is required to expose the `$TELEPORT_REPLICA_NAME` environment variable to each pod in order to the backend storage be able to write the identity separately. The values for `$TELEPORT_REPLICA_NAME` are dependent on the object type:
+
+- Deployment: `TELEPORT_REPLICA_NAME` is a constant string `{{ .Release.Name }}`.
+- Statefulset: `TELEPORT_REPLICA_NAME` is a dynamic value provided by Kubernetes `fieldPath: metadata.name`.
+
+If for some reason the operator increase the number of replicas after deploying the agent with a Deployment object, all agents will use the exact same identity when accessing the cluster.
+
+### Secret creation and lifecycle
+
+In order to use the secret storage, the secret can be created by [Helm chart](https://github.com/gravitational/teleport/tree/master/examples/chart/teleport-kube-agent) at install time. This is a requirement for the cleanup procedures once the user run `helm uninstall {release}`.
```

Comment:
Hmm I'm not sure why do we need to include the Secret with the Helm chart? The agent would save its identity in the Secret itself after getting it from the auth server, right?

Reply:
Yes, initially I thought about that but once the helm chart is uninstalled we will leave the orphan secrets in the cluster. If the user wants to do a fresh install because the cluster changed, he must remove it manually (we must print errors).

Reply:
We can use Helm post-delete hook to clean up secrets if needed.